swarm.js

```diff
@@ -26,57 +39,100 @@ class Swarm extends EventEmitter {
         const info = peerInfo(null)
         info.connected(socket, isTCP)
         this.emit('connection', socket, info)
+        this.serverSockets += 1
+        this[kIncrPeerCount]()
+        socket.once('close', () => {
+          this.serverSockets -= 1
+          this.emit('disconnection', socket, info)
+          this[kDecrPeerCount]()
+        })
       },
       close: () => this.emit('close')
     })
-    queue.on('readable', this._drain(queue))
+
+    network.tcp.maxConnections = maxServerSockets
+    network.utp.maxConnections = maxServerSockets
+
+    queue.on('readable', this[kDrain](queue))
+
+    this.destroyed = false
+    this.clientSockets = 0
+    this.serverSockets = 0
     this.peers = 0
+
     this.maxPeers = maxPeers
-    this.emphemeral = ephemeral !== false
+    this.maxServerSockets = maxServerSockets
+    this.maxClientSockets = maxClientSockets
+
+    this.open = this.peers < this.maxPeers
+    this.ephemeral = ephemeral !== false
+
     this.network = network
-    this.queue = queue
+    this[kQueue] = queue
   }
-  _drain (queue) {
+  [kDrain] (queue) {
     const onConnect = (info) => (err, socket, isTCP) => {
       if (err) {
-        this.peers -= 1
+        this.clientSockets -= 1
+        this[kDecrPeerCount]()
         queue.requeue(info)
         drain()
         return
       }
       info.connected(socket, isTCP)
       this.emit('connection', socket, info)
       socket.on('close', () => {
-        this.peers -= 1
+        this.clientSockets -= 1
+        this.emit('disconnection', socket, info)
+        this[kDecrPeerCount]()
         info.disconnected()
         queue.requeue(info)
+        setImmediate(drain)
       })
       drain()
     }
     const drain = () => {
-      if (this.peers >= this.maxPeers) return
+      if (this.clientSockets >= this.maxClientSockets) return
```

Comment:
i think `|| !this.open` is needed as well here

Reply:
👍

Reply:
👍